### PR TITLE
Change the LogSink interface to take an optional Context object

### PR DESCRIPTION
**Pull request by darkgnotic**
Logger-side changes to support https://github.com/rocicorp/mono/issues/191

Also rename `LogContext.addContext()` to `LogContext.withContext()` to better convey the immutability of the API.

Note: breaking API change

### CODE REVIEW

**Comment by d-llama** [comment]
Adding method name change. Will update this PR shortly.

**Comment by darkgnotic** [comment]
FYI, I will wait for review on this. (No rush, I know that there are a lot of things happening)

File: src/logger.ts [hunk 1]
```diff
@@ -98,17 +109,22 @@ export class FormatLogger implements LogSink {
     this._format = format;
   }
 
-  log(level: LogLevel, ...args: unknown[]): void {
-    console[level](...this._format(level, ...args));
+  log(level: LogLevel, context: Context | undefined, ...args: unknown[]): void {
+    console[level](
+      ...this._format(level, ...stringified(context).concat(args)),
+    );
   }
```
Comment (arv) [inline src/logger.ts]:
```suggestion
    console[level](
      ...this._format(
        level,
        ...stringified(context),
        ...args),
    );
  }
```

File: src/logger.ts [hunk 2]
```diff
@@ -98,17 +109,22 @@ export class FormatLogger implements LogSink {
...
 /**
  * A console logger that prefixes log lines with a log level.
  */
 export const nodeConsoleLogSink: LogSink = {
-  log(level: LogLevel, ...args: unknown[]): void {
-    console[level](logLevelPrefix[level], ...args.map(normalizeArgument));
+  log(level: LogLevel, context: Context | undefined, ...args: unknown[]): void {
+    console[level](
+      logLevelPrefix[level],
+      ...stringified(context).concat(args).map(normalizeArgument),
```
Comment (arv) [inline src/logger.ts]: I do not think you need to call `normalizeArgument` on the stringified context

File: src/logger.ts [hunk 3]
```diff
@@ -128,39 +144,49 @@ export const logLevelPrefix = {
...
+ *   const lc2 = lc.withContext('foo');
+ *   f(lc2);  // logging inside f will contain 'foo' in the Context.
  */
 export class LogContext extends OptionalLoggerImpl {
   private readonly _logSink: LogSink;
   private readonly _level: LogLevel;
-
-  constructor(level: LogLevel = 'info', logSink: LogSink = consoleLogSink) {
-    super(logSink, level);
+  private readonly _context?: Context;
```
Comment (arv) [inline src/logger.ts]: no optional

File: src/logger.test.ts [hunk 4]
```diff
@@ -164,20 +165,22 @@ test('Optional tag', () => {
...
 
-  const lc3 = lc.addContext('d', 'e');
+  const lc3 = lc.withContext('d', 'e');
   lc3.debug?.('f');
   expect(mockDebug.lastCall.args).to.deep.equal(['d=e', 'f']);
 });
 
 class TestLogSink implements LogSink {
-  messages: [LogLevel, ...unknown[]][] = [];
+  // Note: Order in messages is different from log args order to verify
```
Comment (arv) [inline src/logger.test.ts]: You could also have used:

```ts
  messages: [LogLevel, Context | undefined, unknown[]][] = [];
```

File: src/logger.test.ts [hunk 5]
```diff
@@ -194,34 +197,57 @@ test('TeeLogSink', () => {
...
+  lc.withContext('foo', {bar: 'baz'}).debug?.(3, 4);
+  lc.withContext('boo', 'oof').info?.(5, 6);
+  lc.withContext('abc', 'is').withContext('easy', 'as').info?.(1, 2, 3);
+  lc.debug?.(7, 8);
+
+  expect(sink.messages).to.deep.equal([
+    [undefined, 'info', 1, 2],
+    [{foo: {bar: 'baz'}}, 'debug', 3, 4],
+    [{boo: 'oof'}, 'info', 5, 6],
+    [{abc: 'is', easy: 'as'}, 'info', 1, 2, 3],
```
Comment (arv) [inline src/logger.test.ts]: :heart:

**Review by arv — APPROVED** [review]

**Comment by arv** [comment]
LGTM

File: src/logger.test.ts (hunk 4 above)
Comment (darkgnotic) [reply src/logger.test.ts]: Love it. Now my verification code reads as it should:

```js
    ['info', {abc: 'is', easy: 'as'}, [1, 2, 3]],
```

Thanks!

File: src/logger.test.ts (hunk 5 above)
Comment (darkgnotic) [reply src/logger.test.ts]: 😁

File: src/logger.ts (hunk 1 above)
Comment (darkgnotic) [reply src/logger.ts]: Excellent, thank you!

File: src/logger.ts (hunk 2 above)
Comment (darkgnotic) [reply src/logger.ts]: Good point, then I can unroll this one too.

File: src/logger.ts (hunk 3 above)
Comment (darkgnotic) [reply src/logger.ts]: Done.

**Review by darkgnotic — COMMENTED** [review]
Thank you @arv!

**Pull request merged**